client/src/components/uploadProfile/index.tsx

```diff
@@ -0,0 +1,189 @@
+import axios from 'axios';
+
+import { FHIR_API_URL } from '../../constants';
+import { validate } from './validate';
+
+import './style.scss';
+
+interface Resource {
+  id: string;
+  name: string;
+  type: string;
+}
+
+interface Props {
+  isOpen: boolean;
+  resource?: Resource;
+  onClose: () => void;
+  onUpload: (r: Resource) => void;
+}
+
+const overviewAttributes = [
+  'id',
+  'name',
+  'description',
+  'fhirVersion',
+  'type',
+  'publisher'
+];
+const findMissingAttributes = (profile: any) =>
+  overviewAttributes.filter(a => !Object.keys(profile).includes(a));
+
+const formatErrors = (errors: ErrorObject[]): string[] =>
+  errors.map(err => `${err.message}: ${Object.values(err.params).join(',')}`);
+
+const UploadProfile = ({ isOpen, resource, onClose, onUpload }: Props) => {
+  const {
+    getRootProps,
+    getInputProps,
+    isDragActive,
+    acceptedFiles,
+    rejectedFiles
+  } = useDropzone({
+    multiple: false,
+    accept: 'application/json'
+  });
+  const [profileDefinition, setProfileDefinition] = React.useState(
+    undefined as any
+  );
+  const [invalidProfileErrors, setinvalidProfileErrors] = React.useState(
+    undefined as string[] | undefined
+  );
+  const [uploadingProfile, setUploadingProfile] = React.useState(false);
+
+  const validateProfile = (profile: any): boolean => {
+    if (!validate(profile)) {
+      const formattedErrors = formatErrors(validate.errors!);
+      setinvalidProfileErrors(formattedErrors);
+      return false;
+    }
+
+    const missingAttributes = findMissingAttributes(profile);
+    if (missingAttributes.length > 0) {
+      setinvalidProfileErrors(
+        missingAttributes.map(a => `Missing required attribute ${a}`)
+      );
+      return false;
+    }
+    if (profile.type !== resource!.type) {
+      setinvalidProfileErrors([
+        `Cannot create profile of type "${profile.type}" on resource ${
+          resource!.type
+        }`
+      ]);
+      return false;
+    }
+
+    return true;
+  };
+
+  const uploadProfile = async () => {
+    setUploadingProfile(true);
+    try {
+      await axios.post(
+        `${FHIR_API_URL}/StructureDefinition`,
+        profileDefinition
+      );
+
+      onUpload(profileDefinition);
+      onClose();
+    } catch (err) {
+      setinvalidProfileErrors([err.response ? err.response.data : err.message]);
+    }
+    setUploadingProfile(false);
+  };
+
+  React.useEffect(() => {
+    if (!acceptedFiles.length) return;
+
+    const [profileFile] = acceptedFiles;
+    const reader = new FileReader();
+
+    reader.onerror = () =>
+      setinvalidProfileErrors(['could not read selected file']);
+    reader.onload = () => {
+      const profile = JSON.parse(reader.result as string);
+      if (validateProfile(profile)) {
+        setProfileDefinition(profile);
+        setinvalidProfileErrors(undefined);
+      }
+    };
+    reader.readAsText(profileFile);
+  }, [acceptedFiles]);
+
+  const renderError = () => {
```

Comment:
I don't think it's needed to have a function here

Reply:
done.